### PR TITLE
Added focus parent for v7

### PR DESCRIPTION
**Pull request by fhorinek**
Version modified to work with v7 
Old discussion #1503

### CODE REVIEW

**Comment by embeddedt** [comment]
Checks failed: ` error: "LV_USE_GROUP_FOCUS_PARENT" is not defined, evaluates to 0 [-Werror=undef]`

I think you need to run `scripts/lv_conf_checker.py`.

**Comment by fhorinek** [comment]
should work now :-)

**Review by embeddedt — APPROVED** [review]

**Comment by kisvegabor** [comment]
Thank you for rebasing to v7. 

One thing is nor clear to me. The goal is to focus on an arbitrary object or only the direct parent?

**Comment by fhorinek** [comment]
the goal is to have uniform menu layout like this 
image
In general it should be direct parent, but sometimes when the control or layout of the entry will be more complex, the focus target might be parent of parent or even higher in the hierarchy. 
Therefore I used the direct reference instead of the use parent switch.

**Comment by kisvegabor** [comment]
I see, thanks. So it's "only" a visual clue to highlight a parent or grandparent, etc? 

If so, it'd be more lightweight to create something like the [drag_parent](https://github.com/lvgl/lvgl/blob/master/src/lv_core/lv_obj.h#L218) property. It can be cascaded to finally focus the grandparent too. 
It takes only 1 bit and can be added to `lv_obj_t` instead of a `group`. In v7 focusing works without groups too. So if you press a button with a touchpad, the button will be focused. Similarly to desktop apps.

With the `focus_parent` bit we have the option to bring all the affected object to focused or edited state (i.e. the object itself, its parent and grandparent).

What do you think?

**Comment by fhorinek** [comment]
Ok I can rework it, I guess that I do not need a enable macro for one bit

**Comment by fhorinek** [comment]
I have changed the behavior to use parent_focus, however it works only for the groups, since I have no idea how the clicking works. 
My hook where I override the behavior is  `lv_obj_signal` but it turns out that it is not working for mouse/touch interaction

**Comment by kisvegabor** [comment]
I've tested your code with click focus and seems working. I've set `focus_parent` for the switch.

fo

File: src/lv_core/lv_obj.c [hunk 1]
```diff
@@ -3565,6 +3602,23 @@ static lv_design_res_t lv_obj_design(lv_obj_t * obj, const lv_area_t * clip_area
     return LV_DESIGN_RES_OK;
 }
 
+
+/**
+ * Get the really dragged object by taking `focus_parent` into account.
```
Comment (kisvegabor) [inline src/lv_core/lv_obj.c]:
```suggestion
 * Get the really focused object by taking `focus_parent` into account.
```

File: src/lv_core/lv_group.c [hunk 2]
```diff
@@ -532,4 +542,33 @@ static void obj_to_foreground(lv_obj_t * obj)
...
+ * @param child Child object
+ * @param parent Parent object
+ */
+void lv_group_set_focus_parent(lv_obj_t * child, lv_obj_t * parent)
+{
+    child->group_focus_parent = parent;
+
+    /*if the child is in focused state right now, set focus state to parent*/
+    if(lv_obj_is_focused(child)) {
+        lv_obj_clear_state(child, LV_STATE_FOCUSED | LV_STATE_EDITED);
```
[the remark this answers is not in the document]
Comment (fhorinek) [reply src/lv_core/lv_group.c]: this is not to exit the edit mode, just to cancel the state, so it can be transferred to parent, also old commit

**Comment by kisvegabor** [comment]
All looks good. I've added the `next minor` label so it will be added to the next minor release.

When v7 was released I updated out policy about [contributing](https://github.com/lvgl/lvgl/blob/master/docs/CONTRIBUTING.md#how-can-i-send-fixes-and-improvements). To keep the documentation up to date the new features should be added to the docs too. It means only adding a few lines in [lv_obj.md](https://github.com/lvgl/docs/blob/master/v7/en/widgets/obj.md) about it. Could you do it?

**Comment by fhorinek** [comment]
looks like that the group focus and mouse focus is working independently.
Is pointer using different method to focus than `lv_obj_signal`

focus

BTW what are you using to create such nice gifs?

**Comment by embeddedt** [comment]
I don't know what @kisvegabor uses, but I've found that [Peek](https://github.com/phw/peek) is a great tool for doing GIF recordings on Linux ([Ubuntu PPA](https://github.com/phw/peek#ubuntu)).

**Comment by kisvegabor** [comment]
@fhorinek I pushed a fix to handle focus_parent with pointer indevs.

I also use Peek to create GIFs.

**Comment by kisvegabor** [comment]
I've updated our [release policy](https://github.com/lvgl/lvgl#release-policy) and created the `dev` branch to hold new features until they are merged into `master`.

Thank you very much for your contribution!

**Pull request merged**